File: Python/day11/main.py
Point = dict[str, int]


def load_data(filename: str) -> list[Point]:
    galaxies: list[Point] = []
    with open(filename, 'r') as file:
        for y, line in enumerate(file):
            for x, char in enumerate(line):
                if char == "#":
                    galaxies.append({"x": x, "y": y})
    return galaxies


def spread_galaxies(galaxies: list[Point], spread_factor: int):
    for coord in ["x", "y"]:
        galaxies.sort(key=lambda g: g[coord])
        current_spread = 0
        curren_pos = 0
        for galaxy in galaxies:
            diff = galaxy[coord] - curren_pos
            curren_pos = galaxy[coord]
            if diff > 1:
                current_spread += (diff - 1) * (spread_factor - 1)
            galaxy[coord] += current_spread


def solution(galaxies: list[Point], spread_factor: int):
    spread_galaxies(galaxies, spread_factor)
    total_distance = sum_distance(galaxies)
    print(total_distance)


def sum_distance(galaxies: list[Point]) -> int:
    total_distance = 0
    for i in range(len(galaxies)):
        for j in range(i + 1, len(galaxies)):
            total_distance += abs(galaxies[i]["x"] - galaxies[j]["x"]) + abs(galaxies[i]["y"] - galaxies[j]["y"])
    return total_distance


def solution_part_1():
    galaxies = load_data("input.txt")
    solution(galaxies, 2)


def solution_part_2():
    galaxies = load_data("input.txt")
    solution(galaxies, 1_000_000)


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()

File: golang/day11/main.go
package main

import (
	"bufio"
	"os"
	"sort"
)

type Point = map[string]int

func loadData(filename string) ([]Point, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func(data *os.File) {
		err := data.Close()
		if err != nil {
			panic(err)
		}
	}(data)
	var galaxies []Point
	scanner := bufio.NewScanner(data)
	y := 0
	for scanner.Scan() {
		for x, char := range scanner.Text() {
			if char == '#' {
				galaxies = append(galaxies, Point{"x": x, "y": y})
			}
		}
		y += 1
	}
	return galaxies, nil
}

func spreadGalaxies(galaxies []Point, spreadFactor int) {
	for _, coord := range []string{"x", "y"} {
		sort.Slice(galaxies, func(i, j int) bool {
			return galaxies[i][coord] < galaxies[j][coord]
		})
		currentSpread := 0
		currentPos := 0
		for _, galaxy := range galaxies {
			diff := galaxy[coord] - currentPos
			currentPos = galaxy[coord]
			if diff > 1 {
				currentSpread += (diff - 1) * (spreadFactor - 1)
			}
			galaxy[coord] += currentSpread
		}
	}
}

func abs(num int) int {
	if num < 0 {
		return -num
	}
	return num
}

func sumDistances(galaxies []Point) int {
	sum := 0
	for i := 0; i < len(galaxies); i++ {
		for j := i + 1; j < len(galaxies); j++ {
			sum += abs(galaxies[i]["x"]-galaxies[j]["x"]) + abs(galaxies[i]["y"]-galaxies[j]["y"])
		}
	}
	return sum
}

func solution(filename string, spreadFactor int) {
	galaxies, err := loadData(filename)
	if err != nil {
		panic(err)
	}
	spreadGalaxies(galaxies, spreadFactor)
	totalDistance := sumDistances(galaxies)
	println(totalDistance)
}

func solutionPart1() {
	solution("input.txt", 2)
}

func solutionPart2() {
	solution("input.txt", 1000000)
}

func main() {
	solutionPart1()
	solutionPart2()
}
